rllib/algorithms/sac/torch/sac_torch_learner.py

```diff
@@ -217,39 +217,29 @@ def compute_loss_for_module(
         if self.config.twin_q:
             q_target_twin_next = self.module[module_id]._qf_target_twin_forward_train(
                 q_batch_next
-            )
-            q_target_next = torch.min(q_target_next, q_target_twin_next, dim=-1)
+            )[QF_PREDS]
+            q_target_next = torch.min(q_target_next, q_target_twin_next)
 
         # Compute value function for next state (see eq. (3) in Haarnoja et al. (2018)).
         # Note, we use here the sampled actions in the log probabilities.
         q_target_next -= alpha * logps_next
         # Now mask all Q-values with terminated next states in the targets.
-        q_next_masked = (
-            (1.0 - batch[SampleBatch.TERMINATEDS].float())
-            # If the current experience is the last in the episode we neglect it as
-            # otherwise the next value would be from a new episode's reset observation.
-            # See for more information the `EpisodeReplayBuffer.sample()` with
-            # `batch_length_T > 0`.
-            * (1.0 - batch["is_last"].float())
-            * q_target_next
-        )
+        q_next_masked = (1.0 - batch[SampleBatch.TERMINATEDS].float()) * q_target_next
 
         # Compute the right hand side of the Bellman equation.
         # Detach this node from the computation graph as we do not want to
         # backpropagate through the target network when optimizing the Q loss.
         q_selected_target = (
             batch[SampleBatch.REWARDS]
-            # TODO (simon): Implement n-step adjustment.
-            # + (self.config["gamma"] ** self.config["n_step"]) * q_next_masked
-            + self.config.gamma * q_next_masked
+            + (self.config.gamma**self.config.n_step) * q_next_masked
```

Comment:
Nice!

I see, so supporting a uniform n-step range (uniformly sample from n-step values between x and y) will not be that simple, I guess, b/c of this loss term here. We would have to know for each item in the batch, what the exact n-step was.

Reply:
Exactly. We would need to provide this in `EpisodeReplayBuffer.sample()`. But this is not impossible. In fact it could be applied easily. The thing we need to do is to default it to `n_step`. And if it is sampled it gets overridden.